tools/differ/differ.py

```diff
@@ -0,0 +1,251 @@
+        self._cleanup_data(result)
+        result[self.diff_column] = result.apply(lambda row: 'added' if row['added'] > 0 \
+          and row['deleted'] + row['modified'] + row['same'] == 0 \
+          else 'deleted' if row['deleted'] > 0 and row['added'] + row['modified'] + row['same'] == 0 \
+          else 'modified' if row['deleted'] > 0 or row['added'] > 0 or row['modified'] > 0 \
+          else 'same', axis=1)
+        result = result[column_list]
+        result = result.groupby(
+            [self.variable_column, self.diff_column],
+            observed=True,
+            as_index=False)[self.place_column].agg(lambda x: x.tolist())
+        result['size'] = result.apply(lambda row: len(row[self.place_column]),
+                                      axis=1)
+        result[self.place_column] = result.apply(lambda row: random.sample(
+            row[self.place_column],
+            min(SAMPLE_COUNT, len(row[self.place_column]))),
+                                                 axis=1)
+        summary = result.pivot(
+          index=self.variable_column, columns=self.diff_column, values='size')\
+          .reset_index().rename_axis(None, axis=1)
+        self._cleanup_data(summary)
+        summary['total'] = summary.apply(lambda row: row['added'] + row[
+            'deleted'] + row['modified'] + row['same'],
+                                         axis=1)
+        return summary, result
+
+
+def main(_):
+    '''Runs the differ.'''
+    differ = DatasetDiffer(FLAGS.groupby_columns, FLAGS.value_columns)
+
+    if not os.path.exists(FLAGS.output_location):
+        os.makedirs(FLAGS.output_location)
+    logging.info('Loading data...')
+    current_df = helper.load_data(FLAGS.current_data, FLAGS.output_location)
+    previous_df = helper.load_data(FLAGS.previous_data, FLAGS.output_location)
+
+    logging.info('Processing data...')
+    in_data = differ.process_data(previous_df, current_df)
+
+    logging.info('Point analysis:')
+    summary, result = differ.point_analysis(in_data)
+    result.sort_values(by=[differ.diff_column, differ.variable_column],
+                       inplace=True)
+    print(summary.head(10))
+    print(result.head(10))
+    helper.write_data(summary, FLAGS.output_location,
+                      'point-analysis-summary.csv')
+    helper.write_data(result, FLAGS.output_location,
+                      'point-analysis-results.csv')
+
+    logging.info('Series analysis:')
+    summary, result = differ.series_analysis(in_data)
+    result.sort_values(by=[differ.diff_column, differ.variable_column],
+                       inplace=True)
+    print(summary.head(10))
+    print(result.head(10))
+    helper.write_data(summary, FLAGS.output_location,
+                      'series-analysis-summary.csv')
+    helper.write_data(result, FLAGS.output_location,
+                      'series-analysis-results.csv')
+
+    logging.info('Differ output written to folder: %s', FLAGS.output_location)
```

Comment:
Can this be moved into a function that other modules can directly invoke?

Since it is in the same repo, the import-executor could make a function call instead of a sub-process

Reply:
Moved the logic to helper function run_differ